### PR TITLE
core, core/types, common/optional, state: update BlockData with optional types, add decoding of BlockData to core

### LINKED ISSUE #541: core: Update BlockData to match the spec
In core/types, BlockData has a few commented out variables.

Update BlockData as necessary to conform to the spec (Definition E.2.)

### PR DESCRIPTION
**Pull request by noot**
## Changes
- implement optional.Bytes, optional.Header, optional.Body to represent optional types (which exist in rust but not go)
- implement BlockData type to match spec; also add encoding/decoding functions for BlockData, Block, and Header
- update core.ProcessBlockResponseMessage to decode the Data field of the BlockResponseMessage which represents an array of BlockData

## Tests:
```
go test ./... -short
```

### Issues:
closes #541

### CODE REVIEW

File: core/types/types.go [hunk 1]
```diff
@@ -111,19 +152,303 @@ func (bh *Header) Hash() common.Hash {
...
+			Digest:         [][]byte{}, // TODO: digests are not decoded properly due to SCALE
+		}
+		_, err = sd.Decode(header)
+		if err != nil {
+			return nil, err
+		}
+
+		header.Hash()
+
+		// TODO: fix SCALE :(
```
Comment (ryanchristo) [inline core/types/types.go]: #584

File: core/types/types.go [hunk 2]
```diff
@@ -111,19 +152,303 @@ func (bh *Header) Hash() common.Hash {
...
+		return nil, err
+	}
+
+	if exists == 1 {
+		header := &Header{
+			ParentHash:     common.Hash{},
+			Number:         big.NewInt(0),
+			StateRoot:      common.Hash{},
+			ExtrinsicsRoot: common.Hash{},
+			Digest:         [][]byte{}, // TODO: digests are not decoded properly due to SCALE
```
Comment (ryanchristo) [inline core/types/types.go]: #584

File: core/service.go [hunk 3]
```diff
@@ -250,14 +251,52 @@ func (s *Service) ProcessBlockAnnounceMessage(msg p2p.Message) error {
...
+			}
+
+			err = s.executeBlock(enc)
+			if err != nil {
+				log.Error("Failed to validate block", "err", err)
+				return err
+			}
+		}
+
+		// TODO: set BlockData
```
Comment (ryanchristo) [inline core/service.go]: #541

File: common/optional/types.go [hunk 4]
```diff
@@ -71,11 +116,88 @@ func (x *Hash) Value() common.Hash {
...
+	x.exists = exists
+	x.value = value
+}
+
+// CoreBody is the extrinsics inside a state block
+type CoreBody []byte
+
+// Body represents an optional types.Body.
+// The fields need to be exported since it's JSON encoded by the state service.
+// TODO: when we change the state service's encoding to SCALE, these fields should become unexported.
```
Comment (ryanchristo) [inline common/optional/types.go]: #546

File: core/types/types.go (hunk 1 above)
Comment (noot) [reply core/types/types.go]: I can fix this up if #587 gets merged before this

**Pull request merged**